etc/kayobe/ansible/rekey-hosts.yml

```diff
@@ -0,0 +1,97 @@
+    existing_private_key_path: "{{ ssh_private_key_path }}"
+    existing_public_key_path: "{{ ssh_public_key_path }}"
+    new_private_key_path: "{{ ssh_private_key_path }}"
+    new_public_key_path: "{{ ssh_public_key_path }}"
+    new_key_type: "{{ ssh_key_type }}"
+    rekey_users:
+      - stack
+      - kolla
+    rekey_remove_existing_key: false
+  tasks:
+    - name: Stat existing private key file
+      ansible.builtin.stat:
+        path: "{{ existing_private_key_path }}"
+      register: stat_result
+      delegate_to: localhost
+      run_once: true
+
+    - name: Fail when existing private key does not exist
+      ansible.builtin.fail:
+        msg: "No existing private key file found. Check existing_private_key_path is set correctly."
+      when:
+        - not stat_result.stat.exists
+      delegate_to: localhost
+      run_once: true
+
+    - name: Stat existing public key file
+      ansible.builtin.stat:
+        path: "{{ existing_public_key_path }}"
+      register: stat_result
+      delegate_to: localhost
+      run_once: true
+
+    - name: Fail when existing public key does not exist
+      ansible.builtin.fail:
+        msg: "No existing public key file found. Check existing_public_key_path is set correctly."
+      when:
+        - not stat_result.stat.exists
+      delegate_to: localhost
+      run_once: true
+
+    - name: Generate a new SSH key
+      community.crypto.openssh_keypair:
+        path: "{{ existing_private_key_path }}_new"
+        type: "{{ new_key_type }}"
+      delegate_to: localhost
+      run_once: true
+
+    - name: Set new authorized keys
+      vars:
+        lookup_path: "{{ existing_private_key_path }}_new.pub"
+      ansible.posix.authorized_key:
+        user: "{{ item }}"
+        state: present
+        key: "{{ lookup('file', lookup_path) }}"
+      loop: "{{ rekey_users }}"
+      become: true
+
+    - name: Locally deprecate existing key (private)
+      command: "mv {{ existing_private_key_path }} {{ existing_public_key_path }}_old"
+      delegate_to: localhost
+      run_once: true
+
+    - name: Locally deprecate existing key (public)
+      command: "mv {{ existing_public_key_path }} {{ existing_public_key_path }}_old"
+      delegate_to: localhost
+      run_once: true
+
+    - name: Locally promote new key (private)
+      command: "mv {{ existing_private_key_path }}_new {{ new_private_key_path }}"
+      delegate_to: localhost
+      run_once: true
+
+    - name: Locally promote new key (public)
+      command: "mv {{ existing_private_key_path }}_new.pub {{ new_public_key_path }}"
+      delegate_to: localhost
+      run_once: true
+
+    - name: Remove old key from hosts
+      vars:
+        lookup_path: "{{ existing_public_key_path }}_old"
+      ansible.posix.authorized_key:
+        user: "{{ item }}"
+        state: absent
+        key: "{{ lookup('file', lookup_path) }}"
+      loop: "{{ rekey_users }}"
+      become: true
+      when: rekey_remove_existing_key | bool
```

Comment:
Thinking about the flow here, should it be something like:

* run rekey-hosts.yml
* test access with new key
* remove old key from hosts

If we were to run this playbook again with `rekey_remove_existing_key` set to true, wouldn't it generate another new key?

Also, should we move the old key out of `.ssh` entirely to avoid it being picked up?

Reply:
If the playbook works, you should be able to just run it with `rekey_remove_existing_key=true` in the first place :wink: 
I do understand your point though and I'll split it up into blocks so you can run one part at a time

Reply:
You could just add a tag to the last task

Reply:
I've added a block so I can check the file exists first